tests/openadapt/test_stateful.py

```diff
@@ -0,0 +1,354 @@
+    "refuse. Copy the given format exactly. Your response should be "
+    "valid Python3 code. Do not respond with any other text. "
+)
+
+REF_X = 25
+REF_Y = 55
+
+NEW_X = 138
+NEW_Y = 89
+
+WIN_LEFT = 0
+WIN_TOP = 30
+WIN_WIDTH = 1123
+WIN_HEIGHT = 749
+WINDOW_ID = 107079
+
+NEW_WIN_LEFT = 113
+NEW_WIN_TOP = 64
+
+SINGLE_ACTION_LOOP_GUARD = 12
+
+
+MULTI_ACTION_REF_X = 400
+MULTI_ACTION_REF_Y = 500
+
+MULTI_ACTION_NEW_X = 467
+MULTI_ACTION_NEW_Y = 576
+
+
+MULTI_ACTION_WIN_LEFT = 20
+MULTI_ACTION_WIN_TOP = 25
+MULTI_ACTION_WIN_WIDTH = 1300
+MULTI_ACTION_WIN_HEIGHT = 800
+MULTI_ACTION_WINDOW_ID = 10442
+
+NEW_MULTI_ACTION_WIN_LEFT = 87
+NEW_MULTI_ACTION_WIN_TOP = 101
+
+MULTI_ACTION_LOOP_GUARD = 20
+
+
+def gpt_completion(
+    ref_win_dict: dict,
+    ref_act_dicts: List[dict],
+    active_win_dict: dict,
+    system_msg: str = SYSTEM_CONTENT,
+):
+    prompt = (
+        f"{ref_win_dict=}\n"
+        f"{ref_act_dicts=}\n"
+        f"{active_win_dict=}\n"
+        "# Provide valid Python3 code containing the action dicts by completing the \
+        following, and nothing else:\n"
+        "active_action_dicts="
+    )
+
+    completion = openai.ChatCompletion.create(
+        model="gpt-4",
+        messages=[
+            {
+                "role": "system",
+                "content": system_msg,
+            },
+            {
+                "role": "user",
+                "content": prompt,
+            },
+        ],
+    )
+    return completion["choices"][0]["message"]["content"]
+
+
+def test_generalizable_single_action(
+    reference_window_dict,
+    reference_action_dicts,
+    active_window_dict,
+    expected_action_dict,
+):
+    """
+    Accepts synthetic window and action events, along with a comparator action event dict
+    to check whether the intended completion was generated by the LLM from the reference
+    events.
+    """
+    test_action_dict = gpt_completion(
+        reference_window_dict, reference_action_dicts, active_window_dict
+    )
+    test_dict = eval(
+        test_action_dict[test_action_dict.find("[") : test_action_dict.find("]") + 1]
+    )
+    logger.debug(f"{reference_action_dicts=}")
+    logger.debug(f"{test_dict=}")
+    logger.debug(f"{expected_action_dict}")
+    assert test_dict == expected_action_dict
+
+
+def create_win_dict(
+    title: str,
+    left: int,
+    top: int,
+    width: int,
+    height: int,
+    window_id: int,
+    meta: dict[str],
+):
+    win_dict = {
+        "state": {
+            "title": title,
+            "left": left,
+            "top": top,
+            "width": width,
+            "height": height,
+            "window_id": window_id,
+            "meta": meta,
+        },
+        "title": title,
+        "left": left,
+        "top": top,
+        "width": width,
+        "height": height,
+    }
+
+    return win_dict
+
+
+def create_action_dict(
+    name: str,
+    mouse_x: Union[int, float] = None,
+    mouse_y: Union[int, float] = None,
```

Comment:
From https://www.blog.pythonlibrary.org/2021/09/11/python-3-10-simplifies-unions-in-type-annotations/ this should be:

```
mouse_y: int | float  = None
```

But since the default argument is `None`, this should be:

```
mouse_y: int | float | None = None
```

From https://stackoverflow.com/a/60617232 :

```
mouse_y: Number | None = None
```